SQL compatible Null Handling Part 2 - Processing Layer and Druid-SQL changes

nishantmonu51 opened the pull request:
Part 2 of changes for SQL Compatible Null Handling - 
This PR contains changes to improve handling of null values in druid by treating nulls as missing values which don’t actually exist. This will make it more compatible to SQL standard and help with integration with other existing BI systems which support ODBC/JDBC. (Detailed description in proposal - https://github.com/druid-io/druid/issues/4349)

## Nullability for Aggregators/Metrics

As per the current implementation, most aggregators are coded to replace null values with default e.g. sum treats them as zeroes, min treats them as positive infinity, etc.
To match the new semantics and make aggregators nullable, where if an aggregator encounters only the null values the result will be null value following changes are made -

* Aggrgator/BufferAggregator - Added a new method boolean isNull() which returns false by default. aggregators that support nullability can choose to override this and return true if the aggregated result is null. 
* Added a NullableAggregator/NullableBufferAggregator which is be used as a decorator for existing aggregator implementations to make them nullable. This will return null values if all the values aggregated are null, If any of the aggregated value is non-null, it will have same behavior as the delegate aggregator. 
* If new null behavior is enabled, aggregator factories will decorate the aggregators with NullableAggregator/NullableBufferAggregator. 
* Added a new abstract class NullableAggregatorFactory which helps in reducing duplicate code and decide to switch between nullable and non-nullable aggregators. 
* Cardinality and HyperUnique  aggregators ignore null values and only count non-null values, If all the encountered values are null, the result will be 0. This is different from current behavior where null values are also counted.
* Count aggregator - since count aggregator is not associated with any column and just a count of the number of rows, it has the same behavior as before. 

## Math Expressions Null Handling
* For general calculations like sum, full expression is be considered as null if any of the components is null. 
* StringLiteral now supports null and and expressions containing non-quoted null is parsed as StringLiteral with null 
* Specifying a default value for null is supported by the use of NVL or IF clause to assign default values at query time.

## Filtering on Null values
* SelectorDimFilter currently specifies filtering on null values but the implementation assumes null and empty strings as equivalent. The implementation is changed to consider null and empty string differently. Generation of cache key for selectorDimFilter is also modified to support null.
* InFilter/ExpressionFilter/LikeFilter are also modified to not treat nulls as empty strings. 

## Changes to Druid build-in SQL layer
* NULL and empty strings are treated differently. 
* IS NULL and IS NOT NULL filters against null values instead of empty strings. 
* NVL, IFNULL and COALESCE work as per SQL standard. 
* count now skips null values and works as per SQL standard. 

## Backwards Compatibility 
* **_"druid.generic.useDefaultValueForNull"_** is added as a new config. The default value is set to true for backwards compatibility. 
*  code places in query/ingestion where we converted emptyToNull or nullToEmpty now use NullHandlingHelper which only does the conversion if backwards compatibility is enabled. 
* Storage layer changes are backwards compatible for GenericIndexed/String columns. 
* For Numeric columns new V2 serde implementations are introduced which also supports serde of nullRowsBitmaps. Previous implementations are kept unchanged to read any existing segments. 

## Testing 
* Most of the unit tests that tested nulls are modified to have two branches of expected results to test new/old behavior both. 
* For CI have added new runs to test matrix to test new behavior for nulls. 
* To run the tests with new behavior run **_"-Ddruid.generic.useDefaultValueForNull=false"_**

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/druid-io/druid/5452)
<!-- Reviewable:end -->

b-slim commented:
Reviewed 45 of 167 files at r1.
Review status: 45 of 167 files reviewed at latest revision, 1 unresolved discussion.

---

*[processing/src/main/java/io/druid/guice/GuiceInjectors.java, line 45 at r1](https://reviewable.io:443/reviews/druid-io/druid/5452#-L6bm-ZHWBQ5oh4JNF9d:-L6bm-ZHWBQ5oh4JNF9e:b99b0l3) ([raw file](https://github.com/druid-io/druid/blob/a5b6f9bea97bec9a2547b3019e72fa460b11509d/processing/src/main/java/io/druid/guice/GuiceInjectors.java#L45)):*
> ```Java
>         new PropertiesModule(Arrays.asList("common.runtime.properties", "runtime.properties")),
>         new ConfigModule(),
>         new NullHandlingModule(),
> ```

I thought this was done part of the previous patch? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/druid-io/druid/5452)*
<!-- Sent from Reviewable.io -->

nishantmonu51 commented:
@jon-wei @gianm : Can one of you please review this ?

leventov commented on processing/src/main/java/io/druid/query/DefaultQueryMetrics.java:
```diff
@@ -118,7 +118,9 @@ public void duration(QueryType query)
   @Override
   public void queryId(QueryType query)
   {
+    //CHECKSTYLE.OFF: Regexp
```
Please leave a comment here, that this nullToEmpty() is irrelevant to null handling in the data.

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/query/aggregation/AggregatorFactory.java:
```diff
@@ -36,6 +36,8 @@
  * AggregatorFactory is a strategy (in the terms of Design Patterns) that represents column aggregation, e. g. min,
  * max, sum of metric columns, or cardinality of dimension columns (see {@link
  * io.druid.query.aggregation.cardinality.CardinalityAggregatorFactory}).
+ * Implementations of {@link AggregatorFactory} which needs to Support Nullable Aggregations are encouraged
```
Typo: need

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/query/aggregation/AggregatorUtil.java:
```diff
@@ -169,15 +167,15 @@ public void inspectRuntimeShape(RuntimeShapeInspector inspector)
         {
           final ExprEval exprEval = baseSelector.getObject();
-          return exprEval.isNull();
+          return exprEval == null || exprEval.isNull();
```
Why this change?

leventov commented on processing/src/main/java/io/druid/query/aggregation/DoubleMaxAggregatorFactory.java:
```diff
@@ -52,25 +54,43 @@ public DoubleMaxAggregatorFactory(String name, String fieldName)
   @Override
-  public Object combine(Object lhs, Object rhs)
+  protected BufferAggregator factorizeBuffered(
+      ColumnSelectorFactory metricFactory, ColumnValueSelector selector
```
Please format either the whole signature in one line, or each param on a separate string. Same below in Factories.

nishantmonu51 replied:
Will do. 
Can you add a checkstyle rule that fails when this is not done ?

nishantmonu51 replied:
done.

leventov replied:
No, when I checked the last time there was no way to enforce it with checkstyle.

leventov commented on processing/src/main/java/io/druid/query/aggregation/NullableAggregateCombiner.java:
```diff
@@ -0,0 +1,111 @@
+  {
+    boolean isNotNull = !selector.isNull();
+    if (isNotNull) {
+      if (isNullResult) {
```
Should be
```java
if (isNullResult) {
  isNullResult = false;
  delegate.reset(selector);
} else {
  delegate.fold(selector);
}
```

nishantmonu51 replied:
changed the logic to call reset and fold.

leventov commented on processing/src/main/java/io/druid/query/aggregation/NullableAggregateCombiner.java:
```diff
@@ -0,0 +1,111 @@
+  @Override
+  public float getFloat()
+  {
+    if (isNullResult) {
```
`NullHandling.replaceWithDefault()` is not applicable here?

nishantmonu51 replied:
yeah, since this class is only used when replaceWithDefault is disabled.

leventov replied:
Please leave a comment

leventov commented on processing/src/main/java/io/druid/query/aggregation/NullableAggregator.java:
```diff
@@ -0,0 +1,104 @@
+  @Override
+  public float getFloat()
+  {
+    if (isNullResult) {
```
`NullHandling.replaceWithDefault()` is not applicable here?

nishantmonu51 replied:
yeah, since this class is only used when replaceWithDefault is disabled.

leventov replied:
Please leave a comment

nishantmonu51 replied:
Added a class level comment to NullableAgg/BufferAgg/Combiner that they are only used when null handling is enabled.

leventov commented on processing/src/main/java/io/druid/query/aggregation/NullableAggregator.java:
```diff
@@ -0,0 +1,104 @@
+/**
+ * The result of a NullableAggregator will be null if all the values to be aggregated are null values or no values are aggregated at all.
+ * If any of the value is non-null, the result would be the aggregated value of the delegate aggregator.
+ * Note that the delegate aggregator is not required to perform check for isNull on the columnValueSelector as only non-null values
```
Please make isNull a javadoc ref

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/query/aggregation/NullableAggregateCombiner.java:
```diff
@@ -0,0 +1,111 @@
+/**
+ * The result of a NullableAggregateCombiner will be null if all the values to be combined are null values or no values are combined at all.
+ * If any of the value is non-null, the result would be the value of the delegate combiner.
+ * Note that the delegate combiner is not required to perform check for isNull on the columnValueSelector as only non-null values
```
Please make isNull a javadoc ref

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/query/aggregation/NullableAggregatorFactory.java:
```diff
@@ -0,0 +1,78 @@
+
+/**
+ * abstract class with functionality to wrap aggregator/bufferAggregator/combiner to make them Nullable.
+ * Implementations of {@link AggregatorFactory} which needs to Support Nullable Aggregations are encouraged
```
need

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/query/aggregation/NullableAggregatorFactory.java:
```diff
@@ -0,0 +1,78 @@
+import io.druid.segment.ColumnValueSelector;
+
+/**
+ * abstract class with functionality to wrap aggregator/bufferAggregator/combiner to make them Nullable.
```
Please make javadoc refs

leventov commented on processing/src/main/java/io/druid/query/aggregation/NullableAggregatorFactory.java:
```diff
@@ -0,0 +1,78 @@
+    return NullHandling.replaceWithDefault() ? aggregator : new NullableBufferAggregator(aggregator, selector);
+  }
+
+  protected abstract BufferAggregator factorizeBuffered(
```
Move below `ABSTRACT METHODS BELOW`

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/query/aggregation/NullableAggregatorFactory.java:
```diff
@@ -0,0 +1,78 @@
+ * Implementations of {@link AggregatorFactory} which needs to Support Nullable Aggregations are encouraged
+ * to extend this class.
+ */
+public abstract class NullableAggregatorFactory extends AggregatorFactory
```
`@ExtensionPoint`

nishantmonu51 replied:
added.

leventov commented on processing/src/main/java/io/druid/query/aggregation/NullableAggregatorFactory.java:
```diff
@@ -0,0 +1,78 @@
+    return getMaxIntermediateSize2() + (NullHandling.replaceWithDefault() ? 0 : Byte.BYTES);
+  }
+
+  // ---- ABSTRACT METHODS BELOW ------
```
Please add docs to those methods, for extension writers.

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/query/aggregation/NullableBufferAggregator.java:
```diff
@@ -0,0 +1,114 @@
+ */
+public class NullableBufferAggregator implements BufferAggregator
+{
+  private static final byte IS_NULL_BYTE = (byte) 1;
```
I would switch 1 and 0 values. NULL = 0, NOT_NULL = 1 is more natural, plus comparisons with 0 (done below) are cheaper.

nishantmonu51 replied:
switched.

leventov commented on processing/src/main/java/io/druid/query/aggregation/NullableBufferAggregator.java:
```diff
@@ -0,0 +1,114 @@
+/**
+ * The result of a NullableBufferAggregator will be null if all the values to be aggregated are null values or no values are aggregated at all.
+ * If any of the value is non-null, the result would be the aggregated value of the delegate aggregator.
+ * Note that the delegate aggregator is not required to perform check for isNull on the columnValueSelector as only non-null values
```
Please make a javadoc ref

nishantmonu51 replied:
done.

b-slim commented:
Reviewed 123 of 167 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 29 unresolved discussions, some commit checks failed.

---

*[processing/src/main/java/io/druid/query/aggregation/last/DoubleLastAggregatorFactory.java, line 154 at r2](https://reviewable.io:443/reviews/druid-io/druid/5452#-L7STR6TgqZmCU5RnGvS:-L7STR6TgqZmCU5RnGvT:b-nytsme) ([raw file](https://github.com/druid-io/druid/blob/97f6114190ed8f1de5412160a6dea20b2af08f76/processing/src/main/java/io/druid/query/aggregation/last/DoubleLastAggregatorFactory.java#L154)):*
> ```Java
>             if (pair.lhs >= lastTime) {
>               buf.putLong(position, pair.lhs);
>               buf.putDouble(position + Long.BYTES, pair.rhs);
> ```

why are we changing this? I know it is the same size but not really needed.

---

*[processing/src/main/java/io/druid/query/filter/SelectorDimFilter.java, line 79 at r2](https://reviewable.io:443/reviews/druid-io/druid/5452#-L7SWC65uMp6jBwNQtUs:-L7SWC65uMp6jBwNQtUt:b2nh0j3) ([raw file](https://github.com/druid-io/druid/blob/97f6114190ed8f1de5412160a6dea20b2af08f76/processing/src/main/java/io/druid/query/filter/SelectorDimFilter.java#L79)):*
> ```Java
>     byte[] extractionFnBytes = extractionFn == null ? new byte[0] : extractionFn.getCacheKey();
> 
>     return ByteBuffer.allocate(6 + dimensionBytes.length + valueBytes.length + extractionFnBytes.length)
> ```

5 ?

---

*[sql/src/main/java/io/druid/sql/calcite/planner/Calcites.java, line 108 at r2](https://reviewable.io:443/reviews/druid-io/druid/5452#-L7SZez2PgTxv6I94IWB:-L7SZez2PgTxv6I94IWC:beuedeo) ([raw file](https://github.com/druid-io/druid/blob/97f6114190ed8f1de5412160a6dea20b2af08f76/sql/src/main/java/io/druid/sql/calcite/planner/Calcites.java#L108)):*
> ```Java
>   public static String escapeStringLiteral(final String s)
>   {
>     Preconditions.checkNotNull(s);
> ```

this will throw an exception, is that what should be done? the old code seems to be fine with nulls?

---

*[sql/src/main/java/io/druid/sql/calcite/planner/DruidOperatorTable.java, line 121 at r2](https://reviewable.io:443/reviews/druid-io/druid/5452#-L7SZx3sKskwtPhpkvqh:-L7SZx3sKskwtPhpkvqi:blyu1ri) ([raw file](https://github.com/druid-io/druid/blob/97f6114190ed8f1de5412160a6dea20b2af08f76/sql/src/main/java/io/druid/sql/calcite/planner/DruidOperatorTable.java#L121)):*
> ```Java
>           .add(new UnaryPrefixOperatorConversion(SqlStdOperatorTable.NOT, "!"))
>           .add(new UnaryPrefixOperatorConversion(SqlStdOperatorTable.UNARY_MINUS, "-"))
>           .add(new UnaryFunctionOperatorConversion(SqlStdOperatorTable.IS_NULL, "isnull"))
> ```

why not  == null or != null ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/druid-io/druid/5452)*
<!-- Sent from Reviewable.io -->

nishantmonu51 commented on common/src/main/java/io/druid/math/expr/Expr.java:
```diff
@@ -365,15 +365,21 @@ public ExprEval eval(ObjectBinding bindings)
     // Result of any Binary expressions is null if any of the argument is null.
     // e.g "select null * 2 as c;" or "select null + 1 as c;" will return null as per Standard SQL spec.
-    if (NullHandling.sqlCompatible() && (leftVal.isNull() || rightVal.isNull())) {
+    if (NullHandling.sqlCompatible() && (leftVal.value() == null || rightVal.value() == null)) {
```
Reason is that for String ExprEval isNull method now checks that the expression can be parsed as a valid long/double/float or not instead of checking the nullability. 
Have added a comment on the ExprEval isNull method that explains this change.

nishantmonu51 commented on common/src/main/java/io/druid/math/expr/Evals.java:
```diff
@@ -83,6 +83,6 @@ public static boolean asBoolean(double x)
   public static boolean asBoolean(String x)
   {
-    return !Strings.isNullOrEmpty(x) && Boolean.valueOf(x);
+    return !NullHandling.isNullOrEquivalent(x) && Boolean.valueOf(x);
```
sure, will add a checkstyle rule.

nishantmonu51 replied:
added.

nishantmonu51 commented on common/src/main/java/io/druid/math/expr/ExprEval.java:
```diff
@@ -153,7 +154,7 @@ public final double asDouble()
   {
     private DoubleExprEval(Number value)
```
Added @Nullable to these places, I already did one pass and added @Nullable at places, but its easy to miss the @Nullable annotations, Do you have any easy way that i can use to check the missing @Nullable annotations ?

nishantmonu51 replied:
done.

leventov replied:
IntelliJ usually highlights such places with yellow (e. g. when you return `null` from a method not annotated `@Nullable`, or check for non-null an argument not annotated `@Nullable`)

nishantmonu51 commented on common/src/main/java/io/druid/math/expr/ExprEval.java:
```diff
@@ -127,7 +128,7 @@ public String asString()
 
     private NumericExprEval(Number value)
```
done.

nishantmonu51 commented on common/src/main/java/io/druid/math/expr/ExprEval.java:
```diff
@@ -99,6 +97,9 @@ public Object value()
     return value;
```
done.

nishantmonu51 commented on common/src/main/java/io/druid/math/expr/ExprEval.java:
```diff
@@ -245,36 +246,55 @@ public final ExprType type()
+    @Override
+    public boolean isNull()
+    {
+      return asNumber() == null;
```
done.

nishantmonu51 commented on common/src/main/java/io/druid/math/expr/ExprEval.java:
```diff
@@ -99,6 +97,9 @@ public Object value()
+  /**
+   * returns true if numeric primitive value for this ExprEval is null, otherwise false.
+   */
   public boolean isNull()
```
done.

nishantmonu51 commented:
Review status: 128 of 194 files reviewed at latest revision, 29 unresolved discussions.

---

*[processing/src/main/java/io/druid/query/aggregation/last/DoubleLastAggregatorFactory.java, line 154 at r2](https://reviewable.io:443/reviews/druid-io/druid/5452#-L7STR6TgqZmCU5RnGvS:-L7fq9G2qWVtKYILoDkG:blkwbqj) ([raw file](https://github.com/druid-io/druid/blob/97f6114190ed8f1de5412160a6dea20b2af08f76/processing/src/main/java/io/druid/query/aggregation/last/DoubleLastAggregatorFactory.java#L154)):*
<details><summary><i>Previously, b-slim (Slim) wrote…</i></summary><blockquote>

why are we changing this? I know it is the same size but not really needed.
</blockquote></details>

fixed.

---

*[processing/src/main/java/io/druid/query/filter/SelectorDimFilter.java, line 79 at r2](https://reviewable.io:443/reviews/druid-io/druid/5452#-L7SWC65uMp6jBwNQtUs:-L7fp_H2ekcPqIfdVqWA:b-qssebl) ([raw file](https://github.com/druid-io/druid/blob/97f6114190ed8f1de5412160a6dea20b2af08f76/processing/src/main/java/io/druid/query/filter/SelectorDimFilter.java#L79)):*
<details><summary><i>Previously, b-slim (Slim) wrote…</i></summary><blockquote>

5 ?
</blockquote></details>

fixed.

---

*[sql/src/main/java/io/druid/sql/calcite/planner/Calcites.java, line 108 at r2](https://reviewable.io:443/reviews/druid-io/druid/5452#-L7SZez2PgTxv6I94IWB:-L7fpgR9NJP7uKh0Nkz0:bbxldru) ([raw file](https://github.com/druid-io/druid/blob/97f6114190ed8f1de5412160a6dea20b2af08f76/sql/src/main/java/io/druid/sql/calcite/planner/Calcites.java#L108)):*
<details><summary><i>Previously, b-slim (Slim) wrote…</i></summary><blockquote>

this will throw an exception, is that what should be done? the old code seems to be fine with nulls?
</blockquote></details>

yes, we never expect null here now. 

---

*[sql/src/main/java/io/druid/sql/calcite/planner/DruidOperatorTable.java, line 121 at r2](https://reviewable.io:443/reviews/druid-io/druid/5452#-L7SZx3sKskwtPhpkvqh:-L7fpj9ItnRbqqMdQXpm:b-8t2esw) ([raw file](https://github.com/druid-io/druid/blob/97f6114190ed8f1de5412160a6dea20b2af08f76/sql/src/main/java/io/druid/sql/calcite/planner/DruidOperatorTable.java#L121)):*
<details><summary><i>Previously, b-slim (Slim) wrote…</i></summary><blockquote>

why not  == null or != null ?
</blockquote></details>

==  and !=  are binary operators in SQL which when comparing nulls return false as per sql standard. 
e.g. Select null == null; will return false 
and Select isNull(null) will return true; 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/druid-io/druid/5452)*
<!-- Sent from Reviewable.io -->

nishantmonu51 commented:
@leventov @b-slim fixed review comments.

nishantmonu51 commented:
@b-slim @leventov: Any more comments here ?

leventov commented:
@nishantmonu51 I will continue review of this PR. Could you please continue review of my PR?

nishantmonu51 commented:
sure, will try to finish reviewing it this week.

leventov commented on processing/src/main/java/io/druid/query/aggregation/LongMaxAggregatorFactory.java:
```diff
@@ -72,18 +72,24 @@ public LongMaxAggregatorFactory(String name, String fieldName)
+    return new LongMaxBufferAggregator(selector);
+  }
+
+  private ColumnValueSelector getLongColumnSelector(ColumnSelectorFactory metricFactory)
```
Now this private method is used just in one other method. It could be inlined.

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/query/aggregation/LongMinAggregatorFactory.java:
```diff
@@ -73,18 +73,24 @@ public LongMinAggregatorFactory(String name, String fieldName)
+    return new LongMinBufferAggregator(selector);
+  }
+
+  private ColumnValueSelector getLongColumnSelector(ColumnSelectorFactory metricFactory)
```
Same

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/query/aggregation/LongSumAggregatorFactory.java:
```diff
@@ -71,18 +72,24 @@ public LongSumAggregatorFactory(String name, String fieldName)
+    return new LongSumBufferAggregator(selector);
+  }
+
+  private ColumnValueSelector getLongColumnSelector(ColumnSelectorFactory metricFactory)
```
Same

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/query/aggregation/NullableAggregateCombiner.java:
```diff
@@ -0,0 +1,109 @@
+    if (isNullResult) {
+      reset(selector);
+    } else {
+      delegate.fold(selector);
```
The javadoc of this class says that
```
Note that the delegate combiner is not required to perform check for
{@link NullableAggregateCombiner#isNull} on the columnValueSelector
as only non-null values will be passed to the delegate combiner.
```
It is not true here.

nishantmonu51 replied:
done.

leventov replied:
What did you do?

leventov commented on processing/src/main/java/io/druid/query/aggregation/NullableAggregateCombiner.java:
```diff
@@ -0,0 +1,109 @@
+  public void fold(ColumnValueSelector selector)
+  {
+    if (isNullResult) {
+      reset(selector);
```
It's a little counter-intuitive call. I suggest to inline the logic, and not needlessly resetting `isNullResult = true;` each time along the way.

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/query/aggregation/NullableAggregatorFactory.java:
```diff
@@ -0,0 +1,114 @@
+import io.druid.segment.ColumnValueSelector;
+
+/**
+ * abstract class with functionality to wrap {@link Aggregator}, {@link BufferAggregator} and {@link AggregateCombiner} to make them Nullable.
```
Please start with capital letter

leventov replied:
"Make them Nullable" doesn't provide any more information on top of that could be taken from the class name itself. It would be useful to mention "sql compatibility", or link to github issue, or add a link to some other part of the code, etc.

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/query/aggregation/NullableAggregatorFactory.java:
```diff
@@ -0,0 +1,114 @@
+   * @see AggregateCombiner
+   * @see io.druid.segment.IndexMerger
+   */
+  @SuppressWarnings("unused") // Going to be used when https://github.com/druid-io/druid/projects/2 is complete
```
If #5335 is merged first this line needs to be removed

leventov commented on processing/src/main/java/io/druid/query/aggregation/NullableAggregatorFactory.java:
```diff
@@ -0,0 +1,114 @@
+  /**
+   * Creates an {@link AggregateCombiner} to fold rollup aggregation results from serveral "rows" of different indexes during
+   * index merging. AggregateCombiner implements the same logic as {@link #combine}, with the difference that it uses
+   * {@link io.druid.segment.ColumnValueSelector} and it's subinterfaces to get inputs and implements {@code
```
Unnecessary qualifier, strange that it doesn't fail checkstyle

nishantmonu51 replied:
fixed,

leventov commented on processing/src/main/java/io/druid/query/aggregation/NullableBufferAggregator.java:
```diff
@@ -0,0 +1,114 @@
+  private static final byte IS_NOT_NULL_BYTE = (byte) 1;
+  private final BufferAggregator delegate;
+  private final BaseNullableColumnValueSelector selector;
+
```
Extra space

nishantmonu51 replied:
fixed.

leventov commented on processing/src/main/java/io/druid/query/aggregation/NullableBufferAggregator.java:
```diff
@@ -0,0 +1,114 @@
+public class NullableBufferAggregator implements BufferAggregator
+{
+  private static final byte IS_NULL_BYTE = (byte) 0;
+  private static final byte IS_NOT_NULL_BYTE = (byte) 1;
```
Please add space between static and non-static fields

nishantmonu51 replied:
added

leventov commented on processing/src/main/java/io/druid/query/aggregation/NullableBufferAggregator.java:
```diff
@@ -0,0 +1,114 @@
+/**
+ * The result of a NullableBufferAggregator will be null if all the values to be aggregated are null values or no values are aggregated at all.
+ * If any of the value is non-null, the result would be the aggregated value of the delegate aggregator.
+ * Note that the delegate aggregator is not required to perform check for {@link NullableBufferAggregator#isNull(ByteBuffer, int)} on the columnValueSelector as only non-null values
```
Wrong method link

leventov replied:
Please either write "ColumnValueSelector" (starting with capital) to reference the class, or just "selector" because this is the parameter/field name used in code below.

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/query/aggregation/NullableAggregator.java:
```diff
@@ -0,0 +1,104 @@
+/**
+ * The result of a NullableAggregator will be null if all the values to be aggregated are null values or no values are aggregated at all.
+ * If any of the value is non-null, the result would be the aggregated value of the delegate aggregator.
+ * Note that the delegate aggregator is not required to perform check for {@link NullableAggregator#isNull} on the columnValueSelector as only non-null values
```
Wrong method link. Should link `BaseNullableColumnValueSelector.isNull()`.

leventov replied:
Please either write "ColumnValueSelector" (starting with capital) to reference the class, or just "selector" because this is the parameter/field name used in code below.

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/query/aggregation/NullableAggregateCombiner.java:
```diff
@@ -0,0 +1,109 @@
+/**
+ * The result of a NullableAggregateCombiner will be null if all the values to be combined are null values or no values are combined at all.
+ * If any of the value is non-null, the result would be the value of the delegate combiner.
+ * Note that the delegate combiner is not required to perform check for {@link NullableAggregateCombiner#isNull} on the columnValueSelector as only non-null values
```
Wrong method link. Should link `ColumnValueSelector.isNull()`.

leventov replied:
Please either write "ColumnValueSelector" (starting with capital) to reference the class, or just "selector" because this is the parameter name used in code below.

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/query/aggregation/cardinality/types/DoubleCardinalityAggregatorColumnSelectorStrategy.java:
```diff
@@ -20,23 +20,33 @@
 
-
+/**
+ * if performance of this class appears to be a bottleneck for somebody,
```
Please start with capital letter

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/query/aggregation/cardinality/types/FloatCardinalityAggregatorColumnSelectorStrategy.java:
```diff
@@ -20,22 +20,33 @@
 import io.druid.segment.BaseFloatColumnValueSelector;
 
+/**
+ * if performance of this class appears to be a bottleneck for somebody,
```
Same

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/query/aggregation/cardinality/types/LongCardinalityAggregatorColumnSelectorStrategy.java:
```diff
@@ -20,22 +20,33 @@
 import io.druid.segment.BaseLongColumnValueSelector;
 
+/**
+ * if performance of this class appears to be a bottleneck for somebody,
```
Same

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/query/aggregation/first/FloatFirstAggregatorFactory.java:
```diff
@@ -172,9 +185,10 @@ public Object deserialize(Object object)
+  public Object finalizeComputation(@Nullable Object object)
   {
-    return ((SerializablePair<Long, Float>) object).rhs;
+    return object == null ? object : ((SerializablePair<Long, Float>) object).rhs;
```
Double* variant of this class uses `object == null ? null` that is better

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/query/aggregation/first/LongFirstAggregatorFactory.java:
```diff
@@ -165,9 +178,10 @@ public Object deserialize(Object object)
+  public Object finalizeComputation(@Nullable Object object)
   {
-    return ((SerializablePair<Long, Long>) object).rhs;
+    return object == null ? object : ((SerializablePair<Long, Long>) object).rhs;
```
Same

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/query/aggregation/last/DoubleLastAggregatorFactory.java:
```diff
@@ -136,7 +149,7 @@ public void aggregate(ByteBuffer buf, int position)
             if (pair.lhs >= lastTime) {
               buf.putLong(position, pair.lhs);
-              buf.putDouble(position + Long.BYTES, pair.rhs);
+              buf.putDouble(position + Double.BYTES, pair.rhs);
```
It really skips a long, so this line shouldn't be changed

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/query/aggregation/post/DoubleGreatestPostAggregator.java:
```diff
@@ -81,14 +76,12 @@ public Comparator getComparator()
+    while (fieldsIter.hasNext()) {
+      Number nextVal = ((Number) fieldsIter.next().compute(values));
+      // Ignore NULL values and return the greatest out of non-null values.
+      if (nextVal != null && COMPARATOR.compare(nextVal.doubleValue(), retVal) > 0) {
```
This code performs unnecessary unboxing/boxing cycles. I suggest to use `Comparator.comparingDouble(Number::doubleValue)` as `COMPARATOR`, and special-case `retVal == null`

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/query/aggregation/post/DoubleLeastPostAggregator.java:
```diff
@@ -81,14 +76,12 @@ public Comparator getComparator()
+    while (fieldsIter.hasNext()) {
+      Number nextVal = ((Number) fieldsIter.next().compute(values));
+      // Ignore NULL values and return the greatest out of non-null values.
+      if (nextVal != null && COMPARATOR.compare(nextVal.doubleValue(), retVal) < 0) {
```
Same

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/query/aggregation/post/LongGreatestPostAggregator.java:
```diff
@@ -82,14 +76,12 @@ public Comparator getComparator()
+    while (fieldsIter.hasNext()) {
+      Number nextVal = ((Number) fieldsIter.next().compute(values));
+      // Ignore NULL values and return the greatest out of non-null values.
+      if (nextVal != null && COMPARATOR.compare(nextVal.longValue(), retVal) > 0) {
```
Same

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/query/aggregation/post/LongLeastPostAggregator.java:
```diff
@@ -82,14 +76,12 @@ public Comparator getComparator()
+    while (fieldsIter.hasNext()) {
+      Number nextVal = ((Number) fieldsIter.next().compute(values));
+      // Ignore NULL values and return the greatest out of non-null values.
+      if (nextVal != null && COMPARATOR.compare(nextVal.longValue(), retVal) < 0) {
```
Same

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/query/aggregation/post/DoubleLeastPostAggregator.java:
```diff
@@ -37,14 +39,7 @@
-      return ((Double) o).compareTo((Double) o1);
-    }
-  };
+  private static final Comparator COMPARATOR = Ordering.<Double>natural().nullsLast();
```
Please use `Comparator.nullsLast(Comparator.naturalOrder())`. Or add `Comparators.naturalNullsLast()` factory method.

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/query/aggregation/post/LongLeastPostAggregator.java:
```diff
@@ -38,14 +39,7 @@
-      return Longs.compare(((Number) o).longValue(), ((Number) o1).longValue());
-    }
-  };
+  private static final Comparator COMPARATOR = Ordering.<Long>natural().nullsLast();
```
Same

nishantmonu51 replied:
done.

nishantmonu51 commented on common/src/main/java/io/druid/math/expr/ExprEval.java:
```diff
@@ -125,9 +124,9 @@ public String asString()
+    private NumericExprEval(@Nullable Number value)
     {
-      super(value);
+      super(value == null ? NullHandling.defaultDoubleValue() : value);
```
done.

nishantmonu51 commented on common/src/main/java/io/druid/math/expr/ExprEval.java:
```diff
@@ -231,6 +236,9 @@ public Expr toExpr()
 
   private static class StringExprEval extends ExprEval<String>
   {
+    private volatile boolean numericValComputed = false;
```
done.

nishantmonu51 commented on extensions-core/histogram/src/test/java/io/druid/query/aggregation/histogram/sql/QuantileSqlAggregatorTest.java:
```diff
@@ -312,9 +313,14 @@ public void testQuantileOnInnerQuery() throws Exception
-          new Object[]{7.0, 8.26386833190918}
-      );
+      final List<Object[]> expectedResults =
+          NullHandling.replaceWithDefault() ?
```
done.

nishantmonu51 commented on java-util/src/main/java/io/druid/java/util/http/client/NettyHttpClient.java:
```diff
@@ -140,8 +140,10 @@ public void stop()
       channel = channelFuture.getChannel();
     }
-
+    //CHECKSTYLE.OFF: Regexp
```
done.

nishantmonu51 commented on processing/src/main/java/io/druid/query/aggregation/AggregatorUtil.java:
```diff
@@ -156,7 +154,7 @@ public static BaseFloatColumnValueSelector makeColumnValueSelectorWithFloatDefau
         {
           final ExprEval exprEval = baseSelector.getObject();
-          return exprEval.isNull() ? nullValue : (float) exprEval.asDouble();
+          return exprEval.isNumericNull() ? nullValue : (float) exprEval.asDouble();
```
the change in method signature was not required.

nishantmonu51 commented on processing/src/main/java/io/druid/query/aggregation/AggregatorUtil.java:
```diff
@@ -169,15 +167,15 @@ public void inspectRuntimeShape(RuntimeShapeInspector inspector)
   }
 
-  public static BaseLongColumnValueSelector makeColumnValueSelectorWithLongDefault(
+  public static ColumnValueSelector makeColumnValueSelectorWithLongDefault(
```
other change was not required. reverted that.

nishantmonu51 commented on processing/src/main/java/io/druid/query/aggregation/AggregatorUtil.java:
```diff
@@ -210,15 +208,15 @@ public void inspectRuntimeShape(RuntimeShapeInspector inspector)
   }
 
-  public static BaseDoubleColumnValueSelector makeColumnValueSelectorWithDoubleDefault(
+  public static ColumnValueSelector makeColumnValueSelectorWithDoubleDefault(
```
replied.

nishantmonu51 commented:
@leventov: @b-slim: Are you guys done with the review. Any more comments here ?

nishantmonu51 commented on common/src/main/java/io/druid/math/expr/ExprEval.java:
```diff
@@ -126,7 +126,7 @@ public String asString()
 
     private NumericExprEval(@Nullable Number value)
```
It can still be null.

nishantmonu51 commented on indexing-hadoop/src/main/java/io/druid/indexer/path/StaticPathSpec.java:
```diff
@@ -126,10 +126,7 @@ public static void addToMultipleInputs(
-    // String.nulltoempty usage here is irrelevant to null handling of the data.
-    StringBuilder inputFormats = new StringBuilder(Strings.nullToEmpty(conf.get(MultipleInputs.DIR_FORMATS)));
-    //CHECKSTYLE.ON: Regexp
+    StringBuilder inputFormats = new StringBuilder(StringUtils.nullToEmptyNonDruidDataString(conf.get(MultipleInputs.DIR_FORMATS)));
```
formatted. can we make checkstyle fail for such case ?

leventov replied:
We can (I even attempted to make such PR several times), but the amount of code that has to be fixed is enormous so I gave up

nishantmonu51 commented on processing/src/main/java/io/druid/query/aggregation/post/DoubleGreatestPostAggregator.java:
```diff
@@ -80,7 +79,7 @@ public Object compute(Map<String, Object> values)
       // Ignore NULL values and return the greatest out of non-null values.
-      if (nextVal != null && COMPARATOR.compare(nextVal.doubleValue(), retVal) > 0) {
+      if (nextVal != null && COMPARATOR.compare(nextVal, retVal) > 0) {
         retVal = nextVal.doubleValue();
```
done.

nishantmonu51 commented on processing/src/main/java/io/druid/query/aggregation/post/DoubleLeastPostAggregator.java:
```diff
@@ -80,7 +79,7 @@ public Object compute(Map<String, Object> values)
       // Ignore NULL values and return the greatest out of non-null values.
-      if (nextVal != null && COMPARATOR.compare(nextVal.doubleValue(), retVal) < 0) {
+      if (nextVal != null && COMPARATOR.compare(nextVal, retVal) < 0) {
         retVal = nextVal.doubleValue();
```
done.

nishantmonu51 commented on processing/src/main/java/io/druid/query/aggregation/post/LongGreatestPostAggregator.java:
```diff
@@ -80,7 +79,7 @@ public Object compute(Map<String, Object> values)
       // Ignore NULL values and return the greatest out of non-null values.
-      if (nextVal != null && COMPARATOR.compare(nextVal.longValue(), retVal) > 0) {
+      if (nextVal != null && COMPARATOR.compare(nextVal, retVal) > 0) {
         retVal = nextVal.longValue();
```
done.

nishantmonu51 commented on processing/src/main/java/io/druid/query/aggregation/post/LongLeastPostAggregator.java:
```diff
@@ -80,7 +79,7 @@ public Object compute(Map<String, Object> values)
       // Ignore NULL values and return the greatest out of non-null values.
-      if (nextVal != null && COMPARATOR.compare(nextVal.longValue(), retVal) < 0) {
+      if (nextVal != null && COMPARATOR.compare(nextVal, retVal) < 0) {
         retVal = nextVal.longValue();
```
done.

nishantmonu51 commented on processing/src/main/java/io/druid/query/extraction/MapLookupExtractor.java:
```diff
@@ -76,9 +76,7 @@ public String apply(@NotNull String val)
-        //CHECKSTYLE.OFF: Regexp
-        return map.get(key).equals(Strings.nullToEmpty(value));
-        //CHECKSTYLE.ON: Regexp
+        return map.get(key).equals(StringUtils.nullToEmptyNonDruidDataString(value));
```
fixed the lookup extractor. and made null behavior consistent across diff lookupextractor implementations

nishantmonu51 commented on processing/src/main/java/io/druid/query/extraction/StringFormatExtractionFn.java:
```diff
@@ -107,9 +107,7 @@ public String apply(@Nullable String value)
-    //CHECKSTYLE.OFF: Regexp
-    return Strings.emptyToNull(StringUtils.format(format, value));
-    //CHECKSTYLE.ON: Regexp
+    return StringUtils.emptyToNullNonDruidDataString(StringUtils.format(format, value));
```
fixed.

nishantmonu51 commented on processing/src/main/java/io/druid/query/filter/InDimFilter.java:
```diff
@@ -118,9 +117,7 @@ public ExtractionFn getExtractionFn()
-      //CHECKSTYLE.OFF: Regexp
-      valuesBytes[index] = StringUtils.toUtf8(Strings.nullToEmpty(value));
-      //CHECKSTYLE.ON: Regexp
+      valuesBytes[index] = StringUtils.toUtf8(StringUtils.nullToEmptyNonDruidDataString(value));
```
Added a comment and fixed.

nishantmonu51 commented on processing/src/main/java/io/druid/query/filter/InDimFilter.java:
```diff
@@ -214,9 +209,7 @@ public Filter toFilter()
-      //CHECKSTYLE.OFF: Regexp
-      retSet.add(Range.singleton(Strings.nullToEmpty(value)));
-      //CHECKSTYLE.ON: Regexp
+      retSet.add(Range.singleton(StringUtils.nullToEmptyNonDruidDataString(value)));
```
fixed.

nishantmonu51 commented on processing/src/main/java/io/druid/query/filter/SelectorDimFilter.java:
```diff
@@ -190,9 +189,7 @@ public boolean equals(Object o)
-    //CHECKSTYLE.OFF: Regexp
-    retSet.add(Range.singleton(Strings.nullToEmpty(value)));
-    //CHECKSTYLE.ON: Regexp
+    retSet.add(Range.singleton(StringUtils.nullToEmptyNonDruidDataString(value)));
```
value can be null for filters, will add nullable annotation.

nishantmonu51 commented on server/src/test/java/io/druid/query/dimension/LookupDimensionSpecTest.java:
```diff
@@ -162,9 +162,7 @@ public void testGetOutputName()
-      //CHECKSTYLE.OFF: Regexp
-      Assert.assertEquals(Strings.emptyToNull(entry.getValue()), dimensionSpec.getExtractionFn().apply(entry.getKey()));
-      //CHECKSTYLE.ON: Regexp
+      Assert.assertEquals(StringUtils.emptyToNullNonDruidDataString(entry.getValue()), dimensionSpec.getExtractionFn().apply(entry.getKey()));
```
done.

nishantmonu51 commented on server/src/main/java/io/druid/query/dimension/LookupDimensionSpec.java:
```diff
@@ -173,11 +173,7 @@ public boolean mustDecorate()
-    byte[] replaceWithBytes = StringUtils.toUtf8(Strings.nullToEmpty(replaceMissingValueWith));
-    //CHECKSTYLE.ON: Regexp
-
+    byte[] replaceWithBytes = StringUtils.toUtf8(StringUtils.nullToEmptyNonDruidDataString(replaceMissingValueWith));
```
added annotation.

nishantmonu51 commented on processing/src/test/java/io/druid/segment/filter/FilterPartitionTest.java:
```diff
@@ -117,9 +117,7 @@ public Filter toFilter()
       if (extractionFn == null) {
```
added.

nishantmonu51 commented:
@leventov : updated Pr based on your comments.

nishantmonu51 commented:
@b-slim @leventov: In case there are no blocker comments on this PR, can we merge this one when travis is green ?

b-slim commented:
no blocker from my side 👍

leventov commented:
I haven't finished my review.

leventov commented on indexing-service/src/test/java/io/druid/indexing/common/task/AppenderatorDriverRealtimeIndexTaskTest.java:
```diff
@@ -222,7 +223,8 @@ public InputRow nextRow()
     public Runnable commit()
     {
-      return () -> {};
+      return () -> {
```
This is `Runnables.getNoopRunnable()`

leventov commented on extensions-core/lookups-cached-single/src/main/java/io/druid/server/lookup/PollingLookup.java:
```diff
@@ -107,8 +108,14 @@ public void close()
   {
+    String keyEquivalent = NullHandling.nullToEmptyIfNeeded(key);
+    if (keyEquivalent == null) {
+      // keyEquivalent is null for SQL Compatible Null Behavior
```
I don't understand this comment (same in other classes)

nishantmonu51 replied:
this is null only in the case of SQL Compatibility, otherwise null will be replaced with empty string in above line.

leventov replied:
Whenever I write that I don't understanding something please edit code or add comments, not just explain in the PR. Future code readers won't see those.

nishantmonu51 replied:
done.

leventov commented on extensions-core/lookups-cached-single/src/main/java/io/druid/server/lookup/PollingLookup.java:
```diff
@@ -129,8 +136,15 @@ public String apply(@NotNull String key)
   {
+    String valueEquivalent = NullHandling.nullToEmptyIfNeeded(value);
+    if (valueEquivalent == null) {
+      // valueEquivalent is null for SQL Compatible Null Behavior
```
Same

nishantmonu51 replied:
added more details.

leventov commented on processing/src/main/java/io/druid/query/extraction/StringFormatExtractionFn.java:
```diff
@@ -107,7 +107,7 @@ public String apply(@Nullable String value)
       }
     }
-    return Strings.emptyToNull(StringUtils.format(format, value));
+    return io.druid.common.config.NullHandling.emptyToNullIfNeeded(StringUtils.format(format, value));
```
`NullHandling` could be imported?

nishantmonu51 replied:
this class already has an enum with same name.

leventov commented on processing/src/main/java/io/druid/query/filter/DruidDoublePredicate.java:
```diff
@@ -19,12 +19,32 @@
 package io.druid.query.filter;
 
-
+@SuppressWarnings("unused")
```
Please add a comment explaining why you had to add this annotation.

nishantmonu51 replied:
I added this as all implementations of DruidDoublePredicate use lambda and intellij was wrongly complaining that the parameter input is unused.

leventov commented on processing/src/main/java/io/druid/query/filter/DruidFloatPredicate.java:
```diff
@@ -22,9 +22,31 @@
 /**
  * FloatPredicate is only supported in Java 8+, so use this to avoid boxing when a float predicate is needed.
  */
+@SuppressWarnings("unused")
```
Same

nishantmonu51 replied:
I added this as all implementations of DruidFloatPredicate use lambda and intellij was wrongly complaining that the parameter input is unused.

leventov commented on processing/src/main/java/io/druid/query/filter/DruidDoublePredicate.java:
```diff
@@ -19,12 +19,32 @@
 
-
+@SuppressWarnings("unused")
 public interface DruidDoublePredicate
```
Is this class an `ExtensionPoint` or not? If yes, please add annotation and Javadocs to methods.

nishantmonu51 replied:
not an extension point.

leventov commented on processing/src/main/java/io/druid/query/filter/DruidFloatPredicate.java:
```diff
@@ -22,9 +22,31 @@
  * FloatPredicate is only supported in Java 8+, so use this to avoid boxing when a float predicate is needed.
  */
+@SuppressWarnings("unused")
 public interface DruidFloatPredicate
```
Is this class an `ExtensionPoint` or not? If yes, please add annotation and Javadocs to methods.

nishantmonu51 replied:
not an extension point.

leventov commented on processing/src/main/java/io/druid/query/filter/DruidLongPredicate.java:
```diff
@@ -28,5 +28,25 @@
 
```
It's strange that you had to add `@SuppressWarnings` in Double and Float variants but not here.

nishantmonu51 replied:
For Long variant there are still implementations that are not lambda expressions, so intellij was not complaining about unused variable.

leventov commented on processing/src/main/java/io/druid/query/filter/DruidLongPredicate.java:
```diff
@@ -28,5 +28,25 @@
 
   DruidLongPredicate ALWAYS_TRUE = input -> true;
 
```
Is this class an `ExtensionPoint` or not? If yes, please add annotation and Javadocs to methods.

nishantmonu51 replied:
not an extension point.

leventov commented on processing/src/main/java/io/druid/query/filter/LongValueMatcherColumnSelectorStrategy.java:
```diff
@@ -32,7 +31,20 @@ public ValueMatcher makeValueMatcher(final BaseLongColumnValueSelector selector,
     final Long matchVal = DimensionHandlerUtils.convertObjectToLong(value);
     if (matchVal == null) {
-      return BooleanValueMatcher.of(false);
+      return new ValueMatcher()
```
Identical code appears in all three variants (Long, Double and Float), could be extracted as an utility ValueMatcher

leventov commented on processing/src/main/java/io/druid/segment/DoubleDimensionHandler.java:
```diff
@@ -99,6 +100,10 @@ public Closeable getSubColumn(Column column)
   @Override
   public Double getEncodedKeyComponentFromColumn(Closeable column, int currRow)
```
This method should be `@Nullable`

nishantmonu51 replied:
method not present now.

leventov commented on processing/src/main/java/io/druid/segment/FloatDimensionHandler.java:
```diff
@@ -99,6 +100,10 @@ public Closeable getSubColumn(Column column)
   @Override
   public Float getEncodedKeyComponentFromColumn(Closeable column, int currRow)
```
And this

nishantmonu51 replied:
method not present now.

leventov commented on processing/src/main/java/io/druid/segment/LongDimensionHandler.java:
```diff
@@ -99,6 +100,10 @@ public Closeable getSubColumn(Column column)
   @Override
   public Long getEncodedKeyComponentFromColumn(Closeable column, int currRow)
```
And this

nishantmonu51 replied:
this is removed in latest code.

leventov commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/RowBasedGrouperHelper.java:
```diff
@@ -936,7 +925,8 @@ private static int compareDimsInRowsWithAggs(
 
   static long estimateStringKeySize(String key)
```
Nullable param

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/RowBasedGrouperHelper.java:
```diff
@@ -1575,20 +1565,36 @@ public boolean putToKeyBuffer(RowBasedKey key, int idx)
       public int getKeyBufferValueSize()
       {
-        return Long.BYTES;
+        return Long.BYTES + Byte.BYTES;
```
Does this mean that new null handling unconditionally adds overhead in groupBy query processing?

nishantmonu51 replied:
yes, we would take more bytes for both the case, kept it this way to avoid having to maintain two separate code branches for each of long/float/double serde helper

nishantmonu51 replied:
done, refactored the null handling, Now we do not have any overhead for non-null case.

leventov commented on processing/src/main/java/io/druid/query/filter/DruidFloatPredicate.java:
```diff
@@ -22,9 +22,32 @@
 /**
  * FloatPredicate is only supported in Java 8+, so use this to avoid boxing when a float predicate is needed.
```
`FloatPredicate` doesn't exist in JDK. Even if it existed, this comment is stale because now the Druid's interface has two methods.

leventov replied:
Please add "Note: this is not a `{@link PublicApi}` or an `{@link ExtensionPoint}` of Druid."

leventov commented on processing/src/main/java/io/druid/query/filter/DruidFloatPredicate.java:
```diff
@@ -22,9 +22,32 @@
 /**
  * FloatPredicate is only supported in Java 8+, so use this to avoid boxing when a float predicate is needed.
  */
+// All implementations are currently lambda expressions and intellij inspections wrongly complains about unused variable.
```
Add a link to https://youtrack.jetbrains.com/issue/IDEA-191743 (when this ticket is fixed, annotation and comment could be removed)

nishantmonu51 replied:
done. removed stale comment and added note.

leventov commented on processing/src/main/java/io/druid/query/filter/DruidDoublePredicate.java:
```diff
@@ -19,12 +19,36 @@
 
-
+/**
+ * DoublePredicate is only supported in Java 8+, so use this to avoid boxing when a double predicate is needed.
```
This comment is stale because now the Druid's interface has two methods.

leventov replied:
Please add "Note: this is not a `{@link PublicApi}` or an `{@link ExtensionPoint}` of Druid."

nishantmonu51 replied:
done. removed stale comment and added note.

leventov commented on processing/src/main/java/io/druid/query/filter/DruidDoublePredicate.java:
```diff
@@ -19,12 +19,36 @@
+/**
+ * DoublePredicate is only supported in Java 8+, so use this to avoid boxing when a double predicate is needed.
+ */
+// All implementations are currently lambda expressions and intellij inspections wrongly complains about unused variable.
```
Add a link to https://youtrack.jetbrains.com/issue/IDEA-191743 (when this ticket is fixed, annotation and comment could be removed)

nishantmonu51 replied:
done. removed stale comment and added note.

leventov commented on processing/src/main/java/io/druid/query/filter/DruidLongPredicate.java:
```diff
@@ -28,5 +28,25 @@
 
```
The class-level Javadoc comment is stale because now the Druid's interface has two methods.

leventov replied:
Please add "Note: this is not a `{@link PublicApi}` or an `{@link ExtensionPoint}` of Druid." to the class-level Javadoc comment.

nishantmonu51 replied:
done. removed stale comment and added note.

nishantmonu51 commented:
@leventov: Updated PR based on your comments, please check.

nishantmonu51 commented:
@leventov: Any more comments here ? 
Can we get this merged if there are no more blockers ?

leventov commented:
@nishantmonu51 I haven't finished my review. I will continue the review soon

leventov commented on extensions-core/lookups-cached-single/src/test/java/io/druid/server/lookup/LoadingLookupTest.java:
```diff
@@ -43,13 +44,20 @@
   {
     Assert.assertEquals(null, loadingLookup.apply(null));
-    Assert.assertEquals(null, loadingLookup.apply(""));
+    if (!NullHandling.replaceWithDefault()) {
```
This is `NullHandling.sqlCompatible()`

nishantmonu51 replied:
fixed.

leventov commented on processing/src/main/java/io/druid/query/aggregation/cardinality/types/StringCardinalityAggregatorColumnSelectorStrategy.java:
```diff
@@ -40,20 +41,34 @@ public void hashRow(DimensionSelector dimSelector, Hasher hasher)
+        // SQL standard spec does not count null values,
+        // Skip counting null values when we are not replacing null with default value.
+        // A special value for null in case null handling is configured to use empty string for null.
+        if (!NullHandling.replaceWithDefault() && !hasNonNullValue && value != null) {
```
`NullHandling.sqlCompatible()`

nishantmonu51 replied:
fixed.

nishantmonu51 commented on processing/src/main/java/io/druid/query/aggregation/LongMaxAggregatorFactory.java:
```diff
@@ -39,7 +39,7 @@
 /**
  */
-public class LongMaxAggregatorFactory extends AggregatorFactory
+public class LongMaxAggregatorFactory extends NullableAggregatorFactory
```
Sure, This can be done, but i would prefer to do it after this PR is merged. I was even exploring more refactorings that can be done to merge the implementations and minimize the duplications.

nishantmonu51 replied:
https://github.com/druid-io/druid/issues/5860

nishantmonu51 replied:
done.

nishantmonu51 commented on processing/src/main/java/io/druid/query/aggregation/SimpleDoubleAggregatorFactory.java:
```diff
@@ -24,16 +24,17 @@
 import java.util.Objects;
 
-public abstract class SimpleDoubleAggregatorFactory extends AggregatorFactory
+public abstract class SimpleDoubleAggregatorFactory extends NullableAggregatorFactory
```
fixed.

nishantmonu51 commented on processing/src/main/java/io/druid/query/aggregation/NullableBufferAggregator.java:
```diff
@@ -0,0 +1,113 @@
+  @Override
+  public double getDouble(ByteBuffer buf, int position)
+  {
+    if (isNull(buf, position)) {
```
fixed.

nishantmonu51 commented on processing/src/main/java/io/druid/query/aggregation/NullableBufferAggregator.java:
```diff
@@ -0,0 +1,113 @@
+  @Override
+  public long getLong(ByteBuffer buf, int position)
+  {
+    if (isNull(buf, position)) {
```
fixed.

nishantmonu51 commented on processing/src/main/java/io/druid/query/aggregation/NullableBufferAggregator.java:
```diff
@@ -0,0 +1,113 @@
+  @Override
+  public float getFloat(ByteBuffer buf, int position)
+  {
+    if (isNull(buf, position)) {
```
fixed.

nishantmonu51 commented on processing/src/main/java/io/druid/query/aggregation/NullableAggregateCombiner.java:
```diff
@@ -0,0 +1,114 @@
+ * Note that the delegate combiner is not required to perform check for {@link BaseNullableColumnValueSelector#isNull()} on the selector as only non-null values
+ * will be passed to the delegate combiner.
+ */
+public class NullableAggregateCombiner implements AggregateCombiner
```
fixed.

nishantmonu51 commented on processing/src/main/java/io/druid/query/aggregation/NullableAggregatorFactory.java:
```diff
@@ -0,0 +1,114 @@
+ * Support Nullable Aggregations are encouraged to extend this class.
+ */
+@ExtensionPoint
+public abstract class NullableAggregatorFactory extends AggregatorFactory
```
fixed.

nishantmonu51 commented on processing/src/main/java/io/druid/query/aggregation/AggregatorUtil.java:
```diff
@@ -222,15 +219,15 @@ public void inspectRuntimeShape(RuntimeShapeInspector inspector)
   }
 
-  public static BaseDoubleColumnValueSelector makeColumnValueSelectorWithDoubleDefault(
+  public static ColumnValueSelector makeColumnValueSelectorWithDoubleDefault(
```
fixed.

nishantmonu51 commented on processing/src/main/java/io/druid/query/aggregation/AggregatorUtil.java:
```diff
@@ -181,15 +178,15 @@ public void inspectRuntimeShape(RuntimeShapeInspector inspector)
   }
 
-  public static BaseLongColumnValueSelector makeColumnValueSelectorWithLongDefault(
+  public static ColumnValueSelector makeColumnValueSelectorWithLongDefault(
```
fixed.

nishantmonu51 commented on processing/src/main/java/io/druid/query/aggregation/AggregatorUtil.java:
```diff
@@ -148,7 +145,7 @@
   }
 
-  public static BaseFloatColumnValueSelector makeColumnValueSelectorWithFloatDefault(
+  public static ColumnValueSelector makeColumnValueSelectorWithFloatDefault(
```
fixed.

nishantmonu51 commented on indexing-service/src/test/java/io/druid/indexing/common/task/AppenderatorDriverRealtimeIndexTaskTest.java:
```diff
@@ -520,7 +523,7 @@ public void testMaxRowsPerSegment() throws Exception
   }
 
-  @Test(timeout = 60_000L)
+  @Test(timeout = 60_00000L)
```
fixed.

nishantmonu51 commented on extensions-core/lookups-cached-single/src/test/java/io/druid/server/lookup/LoadingLookupTest.java:
```diff
@@ -43,13 +44,20 @@
   {
-    Assert.assertEquals(Collections.EMPTY_LIST, loadingLookup.unapply(null));
+    if (NullHandling.sqlCompatible()) {
+      Assert.assertEquals(Collections.EMPTY_LIST, loadingLookup.unapply(null));
```
fixed.

nishantmonu51 commented on extensions-core/lookups-cached-single/src/main/java/io/druid/server/lookup/PollingLookup.java:
```diff
@@ -129,8 +137,16 @@ public String apply(@NotNull String key)
+      // valueEquivalent is null only for SQL Compatible Null Behavior
+      // otherwise null will be replaced with empty string in nullToEmptyIfNeeded above.
+      // null value maps to empty list when SQL Compatible
+      return Collections.EMPTY_LIST;
```
fixed.

nishantmonu51 commented on processing/src/main/java/io/druid/guice/NullHandlingModule.java:
```diff
@@ -0,0 +1,39 @@
+    JsonConfigProvider.bind(binder, "druid.generic", NullValueHandlingConfig.class);
+    binder.requestStaticInjection(NullHandling.class);
+    binder.requestStaticInjection(NullHandling.class);
+
```
fixed.

nishantmonu51 commented on java-util/src/main/java/io/druid/common/config/NullHandling.java:
```diff
@@ -41,12 +41,14 @@
    * See io.druid.guice.NullHandlingModule for details.
    * It does not take effect in all unit tests since we don't use Guice Injection.
-   * For tests default system property is supposed to be used only in tests
+   * default system property is supposed to be used only in tests
```
fixed.

nishantmonu51 commented on indexing-service/src/test/java/io/druid/indexing/common/task/AppenderatorDriverRealtimeIndexTaskTest.java:
```diff
@@ -681,8 +697,8 @@ public void testNoReportParseExceptions() throws Exception
     // Do some queries.
-    Assert.assertEquals(3, sumMetric(task, null, "rows"));
-    Assert.assertEquals(3, sumMetric(task, null, "met1"));
+    Assert.assertEquals(3, sumMetric(task, null, "met1").longValue());
```
fixed.

nishantmonu51 commented on indexing-service/src/test/java/io/druid/indexing/common/task/AppenderatorDriverRealtimeIndexTaskTest.java:
```diff
@@ -492,8 +495,8 @@ public void testMaxRowsPerSegment() throws Exception
     // Do some queries.
-    Assert.assertEquals(2000, sumMetric(task, null, "rows"));
-    Assert.assertEquals(2000, sumMetric(task, null, "met1"));
+    Assert.assertEquals(2000, sumMetric(task, null, "met1").longValue());
```
fixed.

nishantmonu51 commented on extensions-core/lookups-cached-single/src/main/java/io/druid/server/lookup/LoadingLookup.java:
```diff
@@ -131,8 +139,9 @@ public ApplyCallable(String key)
     @Override
     public String call()
     {
+      // When SQL based null handling is disabled,
```
fixed.

nishantmonu51 commented on extensions-core/lookups-cached-global/src/main/java/io/druid/query/lookup/namespace/UriExtractionNamespace.java:
```diff
@@ -396,8 +397,8 @@ public TSVFlatDataParser(
       final DelimitedParser delegate = new DelimitedParser(
-          Strings.emptyToNull(delimiter),
-          Strings.emptyToNull(listDelimiter),
+              StringUtils.emptyToNullNonDruidDataString(delimiter),
```
fixed.

nishantmonu51 commented:
@leventov: Please review. made NullableAggregatorFactory and NullableAggregatorCombiner generic and handled other comments.

nishantmonu51 commented on processing/src/main/java/io/druid/query/aggregation/first/FloatFirstAggregatorFactory.java:
```diff
@@ -46,7 +46,7 @@
 import java.util.Objects;
 
-public class FloatFirstAggregatorFactory extends NullableAggregatorFactory
+public class FloatFirstAggregatorFactory extends NullableAggregatorFactory<ColumnValueSelector>
```
It returns a SerializablePair<Long, Float> instead of simple Float

leventov replied:
The parameter of NullableAggregatorFactory is not the output selector type, it's the input selector type. So it *should* be `BaseFloatColumnValueSelector` for `FloatFirstAggregatorFactory`. The reason why this refactoring couldn't be done right now is that `getCombiningFactory()` returns an anonymous subclass of  `FloatFirstAggregatorFactory`, that IMO wrong. It seems to me that it should be a separate class `FloatFirstCombiningAggregatorFactory extends NullableAggregatorFactory<ColumnValueSelector<SerializablePair<Long, Float>>>`

nishantmonu51 commented on processing/src/main/java/io/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java:
```diff
@@ -46,7 +46,7 @@
 import java.util.Objects;
 
-public class DoubleFirstAggregatorFactory extends NullableAggregatorFactory
+public class DoubleFirstAggregatorFactory extends NullableAggregatorFactory<ColumnValueSelector>
```
It returns a SerializablePair<Long, Double> instead of simple Double

nishantmonu51 commented on processing/src/main/java/io/druid/query/aggregation/last/DoubleLastAggregatorFactory.java:
```diff
@@ -46,7 +46,7 @@
 import java.util.Objects;
 
-public class DoubleLastAggregatorFactory extends NullableAggregatorFactory
+public class DoubleLastAggregatorFactory extends NullableAggregatorFactory<ColumnValueSelector>
```
It returns a Pair instead of primitive object.

nishantmonu51 commented on processing/src/main/java/io/druid/query/aggregation/last/LongLastAggregatorFactory.java:
```diff
@@ -46,7 +46,7 @@
 import java.util.Objects;
 
-public class LongLastAggregatorFactory extends NullableAggregatorFactory
+public class LongLastAggregatorFactory extends NullableAggregatorFactory<ColumnValueSelector>
```
It returns a Pair instead of primitive object

nishantmonu51 commented on processing/src/main/java/io/druid/query/aggregation/last/FloatLastAggregatorFactory.java:
```diff
@@ -46,7 +46,7 @@
 import java.util.Objects;
 
-public class FloatLastAggregatorFactory extends NullableAggregatorFactory
+public class FloatLastAggregatorFactory extends NullableAggregatorFactory<ColumnValueSelector>
```
It returns a Pair instead of primitive object

nishantmonu51 commented on processing/src/main/java/io/druid/query/aggregation/first/LongFirstAggregatorFactory.java:
```diff
@@ -44,7 +44,7 @@
 import java.util.Map;
 
-public class LongFirstAggregatorFactory extends NullableAggregatorFactory
+public class LongFirstAggregatorFactory extends NullableAggregatorFactory<ColumnValueSelector>
```
It returns a Pair instead of primitive object

nishantmonu51 commented on java-util/src/main/java/io/druid/common/config/NullHandling.java:
```diff
@@ -48,8 +48,8 @@
    * It does not take effect in all unit tests since we don't use Guice Injection.
-   * default system property is supposed to be used only in tests
    */
+  // default system property is supposed to be used only in tests
```
actually this comment is not needed at all, will remove.

nishantmonu51 commented on processing/src/main/java/io/druid/query/filter/SelectorDimFilter.java:
```diff
@@ -88,7 +93,7 @@ public SelectorDimFilter(
   public DimFilter optimize()
   {
-    return new InDimFilter(dimension, ImmutableList.of(value), extractionFn).optimize();
+    return new InDimFilter(dimension, Arrays.asList(value), extractionFn).optimize();
```
immutable list does not allow null values.

nishantmonu51 commented on processing/src/main/java/io/druid/query/filter/ValueGetter.java:
```diff
@@ -27,5 +29,6 @@
   // converted to strings. We should also add functions
   // for these and modify ColumnComparisonFilter to handle
   // comparing Long and Float columns to eachother.
+  @Nullable
```
as it returns null when we have numeric dimensions.

leventov replied:
Should be a Javadoc comment

nishantmonu51 commented on processing/src/main/java/io/druid/query/groupby/orderby/DefaultLimitSpec.java:
```diff
@@ -269,7 +269,7 @@ public int compare(Row left, Row right)
   private Ordering<Row> metricOrdering(final String column, final Comparator comparator)
   {
-    return Ordering.from(Comparator.comparing((Row row) -> row.getRaw(column), Comparator.nullsLast(comparator)));
+    return Ordering.from(Comparator.comparing((Row row) -> row.getRaw(column), Comparator.nullsFirst(comparator)));
```
in case of SQL compatibility we need nullsFirst, have modified to not change the behavior for non-sql compatible case to older one.

leventov replied:
Explain this in a code comment

nishantmonu51 commented on processing/src/main/java/io/druid/query/groupby/having/LessThanHavingSpec.java:
```diff
@@ -66,6 +66,10 @@ public void setAggregators(Map<String, AggregatorFactory> aggregators)
+    if (metricVal == null || value == null) {
+      return false;
+    }
     return HavingSpecMetricComparator.compare(row, aggregationName, value, aggregators) < 0;
```
done.

nishantmonu51 commented on processing/src/main/java/io/druid/query/groupby/having/GreaterThanHavingSpec.java:
```diff
@@ -68,6 +68,10 @@ public void setAggregators(Map<String, AggregatorFactory> aggregators)
+    if (metricVal == null || value == null) {
+      return false;
+    }
     return HavingSpecMetricComparator.compare(row, aggregationName, value, aggregators) > 0;
```
done.

nishantmonu51 commented on processing/src/main/java/io/druid/query/groupby/having/EqualToHavingSpec.java:
```diff
@@ -68,6 +68,10 @@ public void setAggregators(Map<String, AggregatorFactory> aggregators)
+    if (metricVal == null || value == null) {
+      return metricVal == null && value == null;
+    }
     return HavingSpecMetricComparator.compare(row, aggregationName, value, aggregators) == 0;
```
done.

nishantmonu51 commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/column/NullableValueGroupByColumnSelectorStrategy.java:
```diff
@@ -0,0 +1,106 @@
+
+  @Override
+  public void processValueFromGroupingKey(
+      GroupByColumnSelectorPlus selectorPlus, ByteBuffer key, Map<String, Object> resultMap, int keyBufferPosition
```
done.

nishantmonu51 commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/column/LongGroupByColumnSelectorStrategy.java:
```diff
@@ -35,10 +37,10 @@ public int getGroupingKeySize()
   @Override
   public void processValueFromGroupingKey(
-      GroupByColumnSelectorPlus selectorPlus, ByteBuffer key, Map<String, Object> resultMap
+      GroupByColumnSelectorPlus selectorPlus, ByteBuffer key, Map<String, Object> resultMap, int keyBufferPosition
```
done.

nishantmonu51 commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/column/GroupByColumnSelectorStrategy.java:
```diff
@@ -59,11 +59,13 @@
    * @param selectorPlus dimension info containing the key offset, value selector, and dimension spec
    * @param resultMap result map for the group by query being served
    * @param key grouping key
+   * @param keyBufferPosition buffer position for the grouping key
```
done.

nishantmonu51 commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/column/FloatGroupByColumnSelectorStrategy.java:
```diff
@@ -35,10 +37,10 @@ public int getGroupingKeySize()
   @Override
   public void processValueFromGroupingKey(
-      GroupByColumnSelectorPlus selectorPlus, ByteBuffer key, Map<String, Object> resultMap
+      GroupByColumnSelectorPlus selectorPlus, ByteBuffer key, Map<String, Object> resultMap, int keyBufferPosition
```
done.

nishantmonu51 commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/column/DoubleGroupByColumnSelectorStrategy.java:
```diff
@@ -35,10 +37,10 @@ public int getGroupingKeySize()
   @Override
   public void processValueFromGroupingKey(
-      GroupByColumnSelectorPlus selectorPlus, ByteBuffer key, Map<String, Object> resultMap
+      GroupByColumnSelectorPlus selectorPlus, ByteBuffer key, Map<String, Object> resultMap, int keyBufferPosition
```
done.

nishantmonu51 commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/RowBasedGrouperHelper.java:
```diff
@@ -1700,6 +1709,77 @@ public BufferComparator getBufferComparator()
+        BufferComparator delegateBufferComparator = this.delegate.getBufferComparator();
+        this.comparator = (lhsBuffer, rhsBuffer, lhsPosition, rhsPosition) -> {
+          boolean isLhsNull = (lhsBuffer.get(lhsPosition + keyBufferPosition) == NullHandling.IS_NULL_BYTE);
+          boolean isrhsNull = (rhsBuffer.get(rhsPosition + keyBufferPosition) == NullHandling.IS_NULL_BYTE);
```
done.

nishantmonu51 commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/RowBasedGrouperHelper.java:
```diff
@@ -1393,12 +1383,31 @@ private RowBasedKeySerdeHelper makeSerdeHelper(
+    )
+    {
+      if (NullHandling.sqlCompatible()) {
+        return new NullableRowBasedKeySerdeHelper(makeNumericSerdeHelper(
```
done.

nishantmonu51 commented on processing/src/main/java/io/druid/query/extraction/MapLookupExtractor.java:
```diff
@@ -64,21 +64,29 @@ public MapLookupExtractor(
+      return Collections.EMPTY_LIST;
+    }
+
+    return Lists.newArrayList(Maps.filterKeys(map, key -> map.get(key).equals(valueToLookup)).keySet());
```
changed to use Streams API

nishantmonu51 commented on processing/src/main/java/io/druid/query/extraction/MapLookupExtractor.java:
```diff
@@ -64,21 +64,29 @@ public MapLookupExtractor(
+      // valueEquivalent is null only for SQL Compatible Null Behavior
+      // otherwise null will be replaced with empty string in nullToEmptyIfNeeded above.
+      // null value maps to empty list when SQL Compatible
+      return Collections.EMPTY_LIST;
```
done.

nishantmonu51 commented on processing/src/main/java/io/druid/query/extraction/MapLookupExtractor.java:
```diff
@@ -64,21 +64,29 @@ public MapLookupExtractor(
-    }).keySet());
+    String valueToLookup = NullHandling.nullToEmptyIfNeeded(value);
+    if (valueToLookup == null) {
+      // valueEquivalent is null only for SQL Compatible Null Behavior
```
done.

nishantmonu51 commented on processing/src/main/java/io/druid/query/extraction/MapLookupExtractor.java:
```diff
@@ -64,21 +64,29 @@ public MapLookupExtractor(
-    return map.get(val);
+    String keyEquivalent = NullHandling.nullToEmptyIfNeeded(key);
+    if (keyEquivalent == null) {
+      // valueEquivalent is null only for SQL Compatible Null Behavior
```
done.

nishantmonu51 commented on processing/src/main/java/io/druid/query/expression/ExprUtils.java:
```diff
@@ -73,7 +74,14 @@ public static PeriodGranularity toPeriodGranularity(
       Chronology chronology = timeZone == null ? ISOChronology.getInstanceUTC() : ISOChronology.getInstance(timeZone);
       final Object value = originArg.eval(bindings).value();
-      origin = value != null ? new DateTime(value, chronology) : null;
+      if (value instanceof String && StringUtils.isEmpty((String) value)) {
```
Would like to do that in a follow up PR soon after. 
Hope its not a blocker for this one.

leventov replied:
1) create an issue to prohibit `StringUtils` (But I actually think this is easier to prohibit it in this PR)
2) It seems to me that usage of `com.google.guava.common.base.Strings.isNullOrEmpty()`, `java.lang.String.isEmpty()`, etc. should be prohibited too in favor of `NullHanlding.isNullOrEquivalent()` and `StringUtils.isNullOrEmptyNonDruidDataString()` (should create such method)

leventov replied:
And fix this line in this PR anyway

nishantmonu51 commented on processing/src/main/java/io/druid/query/aggregation/post/LongLeastPostAggregator.java:
```diff
@@ -82,13 +75,15 @@ public Comparator getComparator()
+    while (fieldsIter.hasNext()) {
+      Number nextVal = ((Number) fieldsIter.next().compute(values));
+      // Ignore NULL values and return the greatest out of non-null values.
+      if (nextVal != null && COMPARATOR.compare(nextVal, retVal) < 0) {
```
added explicit handling for null, the comparator still handles null it seems to be used in groupby queries.

nishantmonu51 commented on processing/src/main/java/io/druid/query/aggregation/post/LongLeastPostAggregator.java:
```diff
@@ -38,14 +38,7 @@
-      return Longs.compare(((Number) o).longValue(), ((Number) o1).longValue());
-    }
-  };
+  private static final Comparator COMPARATOR = Comparator.nullsLast(Comparator.comparingLong(Number::longValue));
```
added explicit handling for null, the comparator still handles null it seems to be used in groupby queries.

leventov replied:
I didn't find how it is used by groupby queries.

nishantmonu51 commented on processing/src/main/java/io/druid/query/aggregation/post/LongGreatestPostAggregator.java:
```diff
@@ -82,13 +75,15 @@ public Comparator getComparator()
+    while (fieldsIter.hasNext()) {
+      Number nextVal = ((Number) fieldsIter.next().compute(values));
+      // Ignore NULL values and return the greatest out of non-null values.
+      if (nextVal != null && COMPARATOR.compare(nextVal, retVal) > 0) {
```
added explicit handling for null, the comparator still handles null it seems to be used in groupby queries.

nishantmonu51 commented on processing/src/main/java/io/druid/query/aggregation/post/LongGreatestPostAggregator.java:
```diff
@@ -38,14 +38,7 @@
-      return Longs.compare(((Number) o).longValue(), ((Number) o1).longValue());
-    }
-  };
+  private static final Comparator COMPARATOR = Comparator.nullsFirst(Comparator.comparingLong(Number::longValue));
```
added explicit handling for null, the comparator still handles null it seems to be used in groupby queries.

nishantmonu51 commented on processing/src/main/java/io/druid/query/aggregation/post/DoubleLeastPostAggregator.java:
```diff
@@ -37,14 +38,7 @@
-      return ((Double) o).compareTo((Double) o1);
-    }
-  };
+  private static final Comparator COMPARATOR = Comparator.nullsLast(Comparator.comparingDouble(Number::doubleValue));
```
added explicit handling for null, the comparator still handles null it seems to be used in groupby queries.

nishantmonu51 commented on processing/src/main/java/io/druid/query/aggregation/post/DoubleLeastPostAggregator.java:
```diff
@@ -81,13 +75,15 @@ public Comparator getComparator()
+    while (fieldsIter.hasNext()) {
+      Number nextVal = ((Number) fieldsIter.next().compute(values));
+      // Ignore NULL values and return the greatest out of non-null values.
+      if (nextVal != null && COMPARATOR.compare(nextVal, retVal) < 0) {
```
added explicit handling for null, the comparator still handles null it seems to be used in groupby queries.

nishantmonu51 commented on processing/src/main/java/io/druid/query/aggregation/post/DoubleGreatestPostAggregator.java:
```diff
@@ -81,13 +75,15 @@ public Comparator getComparator()
+    while (fieldsIter.hasNext()) {
+      Number nextVal = ((Number) fieldsIter.next().compute(values));
+      // Ignore NULL values and return the greatest out of non-null values.
+      if (nextVal != null && COMPARATOR.compare(nextVal, retVal) > 0) {
```
added explicit handling for null, the comparator still handles null it seems to be used in groupby queries.

nishantmonu51 commented on processing/src/main/java/io/druid/query/aggregation/post/DoubleGreatestPostAggregator.java:
```diff
@@ -37,14 +38,7 @@
-      return ((Double) o).compareTo((Double) o1);
-    }
-  };
+  private static final Comparator COMPARATOR = Comparator.nullsFirst(Comparator.comparingDouble(Number::doubleValue));
```
done.

nishantmonu51 commented:
@leventov : handled comments, do you have any more comments ?

nishantmonu51 commented:
@leventov : have you finished your review ? can this be merged ?

leventov commented on processing/src/main/java/io/druid/query/aggregation/post/DoubleGreatestPostAggregator.java:
```diff
@@ -37,14 +38,7 @@
-      return ((Double) o).compareTo((Double) o1);
-    }
-  };
+  private static final Comparator<Number> COMPARATOR = Comparator.nullsFirst(Comparator.comparingDouble(Number::doubleValue));
```
Line longer than 120 cols. Might also be in other similar classes, I didn't check.

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/query/aggregation/NullableAggregator.java:
```diff
@@ -0,0 +1,106 @@
+import javax.annotation.Nullable;
+
+/**
+ * The result of a NullableAggregator will be null if all the values to be aggregated are null values or no values are aggregated at all.
```
120 cols

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/query/aggregation/NullableAggregator.java:
```diff
@@ -0,0 +1,106 @@
+/**
+ * The result of a NullableAggregator will be null if all the values to be aggregated are null values or no values are aggregated at all.
+ * If any of the value is non-null, the result would be the aggregated value of the delegate aggregator.
+ * Note that the delegate aggregator is not required to perform check for {@link BaseNullableColumnValueSelector#isNull()} on the selector as only non-null values
```
120 cols

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/query/aggregation/NullableAggregatorFactory.java:
```diff
@@ -0,0 +1,114 @@
+  );
+
+  /**
+   * Creates an {@link AggregateCombiner} to fold rollup aggregation results from serveral "rows" of different indexes during
```
120 cols

leventov commented:
@nishantmonu51 after addressing latest comments please squash all commits and create a new PR, it's impossible to work with this one anymore.

nishantmonu51 commented:
Closing in favor of https://github.com/apache/incubator-druid/pull/5958
